src/lib.rs

```diff
@@ -585,9 +606,15 @@ impl Corpus {
             .pad_left(self.pad_left.clone())
             .pad_right(self.pad_right.clone())
             .finish();
-        let mut results: Vec<SearchResult> = self
-            .ngrams
-            .values()
+        let mut ngrams_to_consider: HashSet<&Ngram> = HashSet::new();
+        for gram in item.grams.keys() {
+            if let Some(ngrams) =  self.gram_to_ngram.get(gram) {
+                ngrams_to_consider.extend(ngrams.iter());
+            }
+        }
```

Comment:
This might reduce to something like:

```
let ngrams_to_consider: HashSet<&Ngram> = item.grams.keys()
                                                     .filter_map(|g| self.gram_to_ngram.get(g))
                                                     .flatten(|ng| ng.iter())
                                                     .collect();
```

Reply:
Looks nice, but I don't see how to adapt it to the extra step of word->ngram :confused:

Reply:
No worries. It takes some time and practice to wrap your head around iterator combinators. Then you want to solve everything with iterator combinators. Then you start to understand that 1) not everything can be solved through iterator combinators, 2) not everything _should_ be solved with iterator combinators, and 3) some things are just clearer with a for loop (that's not to say that the input to the for loop can't have been run through some combinators first).

```
let ngrams_to_consider: HashSet<&Ngram> = item.grams.keys() // iterator over grams
                                                     .filter_map(|g| self.gram_to_words.get(g)) // iterator over words-vectors
                                                     .flatten(|wv| wv.iter()) // iterator over words
                                                     .filter_map(|w| self.ngrams.get(w)) // iterator over ngrams
                                                     .collect();
```

You're welcome to keep the comments in.  You're also welcome to suggest that there might be too many steps in the chain for clarity's sake, and you'd prefer to keep the for loop.  I'll leave it up to you.

Reply:
Thanks for the explanation of the chain.
I indeed feels like the for loop is simpler in this case, so I lean towards keeping it